### PR TITLE
Implemented Dancer Ability

### PR DESCRIPTION
**Pull request by YounesM**
Adds implementation for "Dancer" ability according to : https://bulbapedia.bulbagarden.net/wiki/Dancer_(Ability).

### CODE REVIEW

**Review by bennybroseph — CHANGES_REQUESTED** [review]
Needs TSDoc documentation on all new classes and functions and code comments where applicable.

**Review by bennybroseph — APPROVED** [review]

**Review by bennybroseph — CHANGES_REQUESTED** [review]
When you say basic implementation, what does that imply?

**Comment by YounesM** [comment]
> When you say basic implementation, what does that imply?

Well it's honeslty not that basic anymore. I tested all the dancing moves and it seems to work as intended in solo and double battles, there might be some weird cases that I didn't think of but think should cover most if not all use cases.

**Review by bennybroseph — CHANGES_REQUESTED** [review]
Could we get a video of this move working in game? 

Looking to get JSDoc documentation as shown here: [Code Comment FAQ](https://discord.com/channels/1125469663833370665/1239957703820836975)

**Comment by YounesM** [comment]
> Could we get a video of this move working in game?

Sure, in a solo battle:

https://github.com/pagefaultgames/pokerogue/assets/11600168/95aac2dc-8998-425c-b419-5ee7f8545dcd

```
-> Player uses self status dance move
-> Opponent copies it
-> Opponent uses damaging dance move
-> Player returns it
```

In double battles :

https://github.com/pagefaultgames/pokerogue/assets/11600168/d398268b-8570-4d1a-bf8f-0c4a6461730c

```
-> Oricorio [Opp. 1] uses Feather Dance on Cyndaquil 
-> Oricorio [Player] ability triggers and uses Feather Dance on the source of the dance (Oricorio Opp. 1)
-> Oricorio [Opp. 2] ability triggers and uses Feather Dance on the same target than Oricorio [Opp. 1] (Cyndaquil)

-> Oricorio [Opp. 2] uses Fiery Dance on Cyndaquil
-> Oricorio [Player] ability triggers and uses Fiery Dance on the source of the dance (Oricorio Opp. 2)
-> Oricorio [Opp. 1] ability triggers and uses Fiery Dance on the same target than Oricorio [Opp. 2] (Cyndaquil)

-> Oricorio [Player] uses Dragon Dance
-> Oricorio [Opp. 1] and [Opp. 2] ability triggers and copies the move on themselves

-> Cyndaquil uses Aqua Step on Oricorio [Opp. 1]
-> Oricorio [Player] ability triggers and uses Aqua Step on the same target than Cyndaquil (Oricorio Opp. 1)
-> Oricorio [Opp. 1] ability triggers and uses Aqua Step on the source of the dance (Cyndaquil)
-> Cyndaquil is KO
-> Oricorio [Opp. 2] ability triggers and uses Aqua Step on the source of the dance (Cyndaquil) but since Cyndaquil is KO the move goes on Oricorio [Player] instead
```

I tried to cover all the possible cases that could happen when a dance move is used and when multiples dancer are in the fight on the video.

File: src/phases.ts [hunk 1]
```diff
@@ -2410,7 +2410,17 @@ export class MovePhase extends BattlePhase {
...
+      if (this.move.getMove().hasFlag(MoveFlags.DANCE_MOVE) && !this.followUp) {
+        // Pokemon with Dancer can be on either side of the battle so we check in both cases
+        this.scene.getPlayerField().forEach(pokemon => {
+          applyPostMoveUsedAbAttrs(PostMoveUsedAbAttr, pokemon, this.move, this.pokemon, this.targets);
+        })
+        this.scene.getEnemyParty().forEach(pokemon => {
+          applyPostMoveUsedAbAttrs(PostMoveUsedAbAttr, pokemon, this.move, this.pokemon, this.targets);
+        })
+      }
+
```
Comment (bennybroseph) [inline src/phases.ts:2423]: Should this occur based on the pokemon's speed?

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/phases.ts (hunk 1 above)
Comment (YounesM) [reply src/phases.ts:2423]: I don't think so (I cannot find something saying it should), I believe that it should just "react" to a dance move and trigger just after a dance move

**Review by bennybroseph — APPROVED** [review]

**Pull request merged**